File: ios-native/Dimo/Auth/WorkOSAPI.swift
import Foundation

struct WorkOSUser: Codable, Sendable, Equatable {
  var id: String
  var email: String
  var firstName: String?
  var lastName: String?
  var profilePictureUrl: String?

  // WorkOS returns snake_case fields; without these keys the names decode as
  // nil and displayName falls back to the email address.
  enum CodingKeys: String, CodingKey {
    case id
    case email
    case firstName = "first_name"
    case lastName = "last_name"
    case profilePictureUrl = "profile_picture_url"
  }

  var displayName: String {
    let parts = [firstName, lastName].compactMap { $0?.trimmingCharacters(in: .whitespaces) }.filter { !$0.isEmpty }
    if !parts.isEmpty { return parts.joined(separator: " ") }
    return email
  }
}

struct WorkOSSession: Sendable {
  var accessToken: String
  var refreshToken: String
  var user: WorkOSUser
  var expiresAt: Date
}

enum WorkOSAPI {
  struct TokenResponse: Decodable {
    var accessToken: String
    var refreshToken: String?
    var user: WorkOSUser

    enum CodingKeys: String, CodingKey {
      case accessToken = "access_token"
      case refreshToken = "refresh_token"
      case user
    }
  }

  static func exchangeCode(
    code: String,
    codeVerifier: String,
    clientId: String,
    redirectURI: String
  ) async throws -> WorkOSSession {
    var request = URLRequest(url: URL(string: "\(AppConfig.workOSAuthBaseURL)/user_management/authenticate")!)
    request.httpMethod = "POST"
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    let body: [String: String] = [
      "client_id": clientId,
      "grant_type": "authorization_code",
      "code": code,
      "code_verifier": codeVerifier,
      "redirect_uri": redirectURI,
    ]
    request.httpBody = try JSONSerialization.data(withJSONObject: body)
    let (data, response) = try await URLSession.shared.data(for: request)
    try throwIfNeeded(response: response, data: data)
    let decoded = try JSONDecoder().decode(TokenResponse.self, from: data)
    guard let refresh = decoded.refreshToken else {
      throw AuthError.missingRefreshToken
    }
    return WorkOSSession(
      accessToken: decoded.accessToken,
      refreshToken: refresh,
      user: decoded.user,
      expiresAt: jwtExpiry(decoded.accessToken) ?? Date().addingTimeInterval(3600)
    )
  }

  /// Exchanges a refresh token. Retries transient transport / 429 / 5xx failures
  /// per WorkOS session-resilience guidance; surfaces `invalid_grant` as terminal.
  static func refresh(
    refreshToken: String,
    clientId: String,
    maxAttempts: Int = 4
  ) async throws -> WorkOSSession {
    var lastError: Error?
    for attempt in 0..<max(1, maxAttempts) {
      do {
        return try await refreshOnce(refreshToken: refreshToken, clientId: clientId)
      } catch let error as AuthError where error.isTerminalRefresh {
        throw error
      } catch {
        lastError = error
        if attempt + 1 >= maxAttempts { break }
        let delay = min(8.0, pow(2.0, Double(attempt))) * (0.75 + Double.random(in: 0...0.5))
        try await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
      }
    }
    throw lastError ?? AuthError.transientRefresh("Refresh failed")
  }

  private static func refreshOnce(
    refreshToken: String,
    clientId: String
  ) async throws -> WorkOSSession {
    var request = URLRequest(url: URL(string: "\(AppConfig.workOSAuthBaseURL)/user_management/authenticate")!)
    request.httpMethod = "POST"
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    request.timeoutInterval = 30
    let body: [String: String] = [
      "client_id": clientId,
      "grant_type": "refresh_token",
      "refresh_token": refreshToken,
    ]
    request.httpBody = try JSONSerialization.data(withJSONObject: body)
    let (data, response) = try await URLSession.shared.data(for: request)
    try throwIfNeeded(response: response, data: data)
    let decoded = try JSONDecoder().decode(TokenResponse.self, from: data)
    // Always prefer the rotated refresh token when WorkOS returns one.
    let refresh = decoded.refreshToken ?? refreshToken
    return WorkOSSession(
      accessToken: decoded.accessToken,
      refreshToken: refresh,
      user: decoded.user,
      expiresAt: jwtExpiry(decoded.accessToken) ?? Date().addingTimeInterval(3600)
    )
  }

  static func authorizationURL(
    clientId: String,
    redirectURI: String,
    state: String,
    codeChallenge: String,
    provider: String
  ) -> URL {
    var components = URLComponents(string: "\(AppConfig.workOSAuthBaseURL)/user_management/authorize")!
    components.queryItems = [
      URLQueryItem(name: "client_id", value: clientId),
      URLQueryItem(name: "redirect_uri", value: redirectURI),
      URLQueryItem(name: "response_type", value: "code"),
      URLQueryItem(name: "provider", value: provider),
      URLQueryItem(name: "state", value: state),
      URLQueryItem(name: "code_challenge", value: codeChallenge),
      URLQueryItem(name: "code_challenge_method", value: "S256"),
    ]
    return components.url!
  }

  static func jwtExpiry(_ token: String) -> Date? {
    let parts = token.split(separator: ".")
    guard parts.count >= 2 else { return nil }
    var payload = String(parts[1])
      .replacingOccurrences(of: "-", with: "+")
      .replacingOccurrences(of: "_", with: "/")
    while payload.count % 4 != 0 { payload += "=" }
    guard let data = Data(base64Encoded: payload),
          let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
      return nil
    }
    let exp: TimeInterval?
    if let value = json["exp"] as? TimeInterval {
      exp = value
    } else if let value = json["exp"] as? Int {
      exp = TimeInterval(value)
    } else if let number = json["exp"] as? NSNumber {
      exp = number.doubleValue
    } else {
      exp = nil
    }
    guard let exp else { return nil }
    return Date(timeIntervalSince1970: exp)
  }

  private static func throwIfNeeded(response: URLResponse, data: Data) throws {
    guard let http = response as? HTTPURLResponse else { return }
    guard (200..<300).contains(http.statusCode) else {
      throw AuthError.fromHTTP(status: http.statusCode, data: data)
    }
  }
}

enum AuthError: LocalizedError {
  case missingRefreshToken
  case cancelled
  case stateMismatch
  case missingCode
  case notAuthenticated
  case server(String)
  /// Refresh token revoked/expired — clear Keychain and send user to sign-in.
  case terminalRefresh(String)
  /// Network / 429 / 5xx — keep Keychain and retry; never treat as sign-out.
  case transientRefresh(String)

  var errorDescription: String? {
    switch self {
    case .missingRefreshToken: return "Missing refresh token"
    case .cancelled: return "Sign-in cancelled"
    case .stateMismatch: return "OAuth state mismatch"
    case .missingCode: return "Missing authorization code"
    case .notAuthenticated: return "Not authenticated"
    case .server(let message): return message
    case .terminalRefresh(let message): return message
    case .transientRefresh(let message): return message
    }
  }

  var isTerminalRefresh: Bool {
    switch self {
    case .terminalRefresh, .notAuthenticated, .missingRefreshToken:
      return true
    default:
      return false
    }
  }

  static func fromHTTP(status: Int, data: Data) -> AuthError {
    let body = String(data: data, encoding: .utf8) ?? ""
    let oauthError = oauthErrorCode(from: data)
    if status == 400, oauthError == "invalid_grant" {
      return .terminalRefresh(sanitizeMessage(body, fallback: "Session expired"))
    }
    if status == 408 || status == 429 || (500...599).contains(status) {
      return .transientRefresh(sanitizeMessage(body, fallback: "HTTP \(status)"))
    }
    // Unexpected auth errors are treated as terminal so we do not loop forever
    // on a permanently rejected client configuration.
    if (400..<500).contains(status) {
      return .terminalRefresh(sanitizeMessage(body, fallback: "HTTP \(status)"))
    }
    return .transientRefresh(sanitizeMessage(body, fallback: "HTTP \(status)"))
  }

  static func isTransient(_ error: Error) -> Bool {
    if let auth = error as? AuthError {
      switch auth {
      case .transientRefresh: return true
      default: return false
      }
    }
    if error is URLError { return true }
    return false
  }

  private static func oauthErrorCode(from data: Data) -> String? {
    guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
      return nil
    }
    if let code = json["error"] as? String { return code }
    if let nested = json["error"] as? [String: Any], let code = nested["code"] as? String {
      return code
    }
    return nil
  }

  private static func sanitizeMessage(_ body: String, fallback: String) -> String {
    let trimmed = body.trimmingCharacters(in: .whitespacesAndNewlines)
    if trimmed.isEmpty { return fallback }
    return String(trimmed.prefix(200))
  }
}

File: android-native/app/src/main/java/app/dimo/android/auth/WorkOSAPI.kt
package app.dimo.android.auth

import android.net.Uri
import android.util.Base64
import app.dimo.android.app.AppConfig
import java.io.IOException
import kotlin.math.min
import kotlin.math.pow
import kotlin.random.Random
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject

/** Port of `ios-native/Dimo/Auth/WorkOSAPI.swift`. */

@Serializable
data class WorkOSUser(
  val id: String,
  val email: String,
  // WorkOS returns snake_case; without these names the fields decode as null and
  // displayName falls back to the email address.
  @SerialName("first_name") val firstName: String? = null,
  @SerialName("last_name") val lastName: String? = null,
  @SerialName("profile_picture_url") val profilePictureUrl: String? = null,
) {
  val displayName: String
    get() {
      val parts = listOfNotNull(firstName, lastName).map { it.trim() }.filter { it.isNotEmpty() }
      return if (parts.isNotEmpty()) parts.joinToString(" ") else email
    }
}

data class WorkOSSession(
  val accessToken: String,
  val refreshToken: String,
  val user: WorkOSUser,
  /** Epoch millis at which [accessToken] expires. */
  val expiresAt: Long,
)

sealed class AuthException(message: String) : Exception(message) {
  data object MissingRefreshToken : AuthException("Missing refresh token")
  data object Cancelled : AuthException("Sign-in cancelled")
  data object StateMismatch : AuthException("OAuth state mismatch")
  data object MissingCode : AuthException("Missing authorization code")
  data object NotAuthenticated : AuthException("Not authenticated")
  data class Server(val detail: String) : AuthException(detail)
  /** Refresh token revoked/expired — clear storage and send user to sign-in. */
  data class TerminalRefresh(val detail: String) : AuthException(detail)
  /** Network / 429 / 5xx — keep storage and retry; never treat as sign-out. */
  data class TransientRefresh(val detail: String) : AuthException(detail)

  val isTerminalRefresh: Boolean
    get() = this is TerminalRefresh || this is NotAuthenticated || this is MissingRefreshToken

  companion object {
    fun fromHttp(status: Int, body: String): AuthException {
      val oauthError = oauthErrorCode(body)
      if (status == 400 && oauthError == "invalid_grant") {
        return TerminalRefresh(sanitize(body, "Session expired"))
      }
      if (status == 408 || status == 429 || status in 500..599) {
        return TransientRefresh(sanitize(body, "HTTP $status"))
      }
      if (status in 400..499) {
        return TerminalRefresh(sanitize(body, "HTTP $status"))
      }
      return TransientRefresh(sanitize(body, "HTTP $status"))
    }

    fun isTransient(error: Throwable): Boolean =
      error is TransientRefresh || error is IOException

    private fun oauthErrorCode(body: String): String? = runCatching {
      val json = JSONObject(body)
      when (val error = json.opt("error")) {
        is String -> error
        is JSONObject -> error.optString("code").ifEmpty { null }
        else -> null
      }
    }.getOrNull()

    private fun sanitize(body: String, fallback: String): String {
      val trimmed = body.trim()
      if (trimmed.isEmpty()) return fallback
      return trimmed.take(200)
    }
  }
}

object WorkOSAPI {
  private val json = Json { ignoreUnknownKeys = true }
  private val jsonMediaType = "application/json".toMediaType()

  private val client: OkHttpClient by lazy {
    OkHttpClient.Builder()
      .callTimeout(30, java.util.concurrent.TimeUnit.SECONDS)
      .build()
  }

  @Serializable
  private data class TokenResponse(
    @SerialName("access_token") val accessToken: String,
    @SerialName("refresh_token") val refreshToken: String? = null,
    val user: WorkOSUser,
  )

  suspend fun exchangeCode(
    code: String,
    codeVerifier: String,
    clientId: String,
    redirectURI: String,
  ): WorkOSSession {
    val body = JSONObject(
      mapOf(
        "client_id" to clientId,
        "grant_type" to "authorization_code",
        "code" to code,
        "code_verifier" to codeVerifier,
        "redirect_uri" to redirectURI,
      ),
    ).toString()
    val decoded = post(body)
    val refresh = decoded.refreshToken ?: throw AuthException.MissingRefreshToken
    return WorkOSSession(
      accessToken = decoded.accessToken,
      refreshToken = refresh,
      user = decoded.user,
      expiresAt = jwtExpiry(decoded.accessToken) ?: (System.currentTimeMillis() + 3_600_000),
    )
  }

  /** Retries transient failures; surfaces terminal `invalid_grant` immediately. */
  suspend fun refresh(
    refreshToken: String,
    clientId: String,
    maxAttempts: Int = 4,
  ): WorkOSSession {
    var lastError: Throwable? = null
    repeat(maxAttempts.coerceAtLeast(1)) { attempt ->
      try {
        return refreshOnce(refreshToken, clientId)
      } catch (error: AuthException) {
        if (error.isTerminalRefresh) throw error
        lastError = error
      } catch (error: IOException) {
        lastError = AuthException.TransientRefresh(error.message ?: "Network error")
      }
      if (attempt + 1 < maxAttempts) {
        val delayMs = (min(8.0, 2.0.pow(attempt.toDouble())) * (0.75 + Random.nextDouble()) * 1000).toLong()
        delay(delayMs)
      }
    }
    throw lastError ?: AuthException.TransientRefresh("Refresh failed")
  }

  private suspend fun refreshOnce(refreshToken: String, clientId: String): WorkOSSession {
    val body = JSONObject(
      mapOf(
        "client_id" to clientId,
        "grant_type" to "refresh_token",
        "refresh_token" to refreshToken,
      ),
    ).toString()
    val decoded = post(body)
    return WorkOSSession(
      accessToken = decoded.accessToken,
      refreshToken = decoded.refreshToken ?: refreshToken,
      user = decoded.user,
      expiresAt = jwtExpiry(decoded.accessToken) ?: (System.currentTimeMillis() + 3_600_000),
    )
  }

  fun authorizationURL(
    clientId: String,
    redirectURI: String,
    state: String,
    codeChallenge: String,
    provider: String,
  ): Uri = Uri.parse("${AppConfig.workOSAuthBaseURL}/user_management/authorize")
    .buildUpon()
    .appendQueryParameter("client_id", clientId)
    .appendQueryParameter("redirect_uri", redirectURI)
    .appendQueryParameter("response_type", "code")
    .appendQueryParameter("provider", provider)
    .appendQueryParameter("state", state)
    .appendQueryParameter("code_challenge", codeChallenge)
    .appendQueryParameter("code_challenge_method", "S256")
    .build()

  /** Epoch millis from the JWT `exp` claim, or null when it cannot be read. */
  fun jwtExpiry(token: String): Long? {
    val parts = token.split(".")
    if (parts.size < 2) return null
    return runCatching {
      val payload = Base64.decode(parts[1], Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
      val exp = JSONObject(String(payload, Charsets.UTF_8)).optDouble("exp", Double.NaN)
      if (exp.isNaN()) null else (exp * 1000).toLong()
    }.getOrNull()
  }

  private suspend fun post(body: String): TokenResponse = withContext(Dispatchers.IO) {
    val request = Request.Builder()
      .url("${AppConfig.workOSAuthBaseURL}/user_management/authenticate")
      .post(body.toRequestBody(jsonMediaType))
      .build()
    try {
      client.newCall(request).execute().use { response ->
        val text = response.body?.string().orEmpty()
        if (!response.isSuccessful) {
          throw AuthException.fromHttp(response.code, text)
        }
        json.decodeFromString(TokenResponse.serializer(), text)
      }
    } catch (io: IOException) {
      throw AuthException.TransientRefresh(io.message ?: "Network error")
    }
  }
}
